// @mui/x-charts/internals/plugins/featurePlugins/useChartKeyboardNavigation/useChartKeyboardNavigation.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useChartKeyboardNavigation = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _useEventCallback = _interopRequireDefault(require("@mui/utils/useEventCallback"));
var _useEnhancedEffect = _interopRequireDefault(require("@mui/utils/useEnhancedEffect"));
var _useChartKeyboardNavigation = require("./useChartKeyboardNavigation.helpers");
function getNextIndexFocusedItem(state) {
  let {
    type,
    seriesId
  } = state.keyboardNavigation.item ?? {};
  if (type === undefined ||
  // @ts-ignore sankey is not in MIT version
  type === 'sankey' || seriesId === undefined || !(0, _useChartKeyboardNavigation.seriesHasData)(state.series.processedSeries, type, seriesId)) {
    const nextSeries = (0, _useChartKeyboardNavigation.getNextSeriesWithData)(state.series.processedSeries, type, seriesId);
    if (nextSeries === null) {
      return null;
    }
    type = nextSeries.type;
    seriesId = nextSeries.seriesId;
  }
  const dataLength = state.series.processedSeries[type].series[seriesId].data.length;
  return {
    type,
    seriesId,
    dataIndex: ((state.keyboardNavigation.item?.dataIndex ?? -1) + 1) % dataLength
  };
}
function getPreviousIndexFocusedItem(state) {
  let {
    type,
    seriesId
  } = state.keyboardNavigation.item ?? {};
  if (type === undefined ||
  // @ts-ignore sankey is not in MIT version
  type === 'sankey' || seriesId === undefined || !(0, _useChartKeyboardNavigation.seriesHasData)(state.series.processedSeries, type, seriesId)) {
    const previousSeries = (0, _useChartKeyboardNavigation.getPreviousSeriesWithData)(state.series.processedSeries, type, seriesId);
    if (previousSeries === null) {
      return null;
    }
    type = previousSeries.type;
    seriesId = previousSeries.seriesId;
  }
  const dataLength = state.series.processedSeries[type].series[seriesId].data.length;
  return {
    type,
    seriesId,
    dataIndex: (dataLength + (state.keyboardNavigation.item?.dataIndex ?? 1) - 1) % dataLength
  };
}
function getNextSeriesFocusedItem(state) {
  let {
    type,
    seriesId
  } = state.keyboardNavigation.item ?? {};
  const nextSeries = (0, _useChartKeyboardNavigation.getNextSeriesWithData)(state.series.processedSeries, type, seriesId);
  if (nextSeries === null) {
    return null; // No series to move the focus to.
  }
  type = nextSeries.type;
  seriesId = nextSeries.seriesId;
  const dataLength = state.series.processedSeries[type].series[seriesId].data.length;
  return {
    type,
    seriesId,
    dataIndex: Math.min(dataLength - 1, state.keyboardNavigation.item?.dataIndex ?? 0)
  };
}
function getPreviousSeriesFocusedItem(state) {
  let {
    type,
    seriesId
  } = state.keyboardNavigation.item ?? {};
  const previousSeries = (0, _useChartKeyboardNavigation.getPreviousSeriesWithData)(state.series.processedSeries, type, seriesId);
  if (previousSeries === null) {
    return null; // No series to move the focus to.
  }
  type = previousSeries.type;
  seriesId = previousSeries.seriesId;
  const dataLength = state.series.processedSeries[type].series[seriesId].data.length;
  return {
    type,
    seriesId,
    dataIndex: Math.min(dataLength - 1, state.keyboardNavigation.item?.dataIndex ?? 0)
  };
}
const useChartKeyboardNavigation = ({
  params,
  store,
  svgRef
}) => {
  const removeFocus = (0, _useEventCallback.default)(function removeFocus() {
    store.update(state => {
      if (state.keyboardNavigation.item === null) {
        return state;
      }
      return (0, _extends2.default)({}, state, {
        keyboardNavigation: (0, _extends2.default)({}, state.keyboardNavigation, {
          item: null
        })
      });
    });
  });
  React.useEffect(() => {
    const element = svgRef.current;
    if (!element || !params.enableKeyboardNavigation) {
      return undefined;
    }
    function keyboardHandler(event) {
      store.update(prevState => {
        let newFocusedItem = prevState.keyboardNavigation.item;
        switch (event.key) {
          case 'ArrowRight':
            newFocusedItem = getNextIndexFocusedItem(prevState);
            break;
          case 'ArrowLeft':
            newFocusedItem = getPreviousIndexFocusedItem(prevState);
            break;
          case 'ArrowDown':
            {
              newFocusedItem = getPreviousSeriesFocusedItem(prevState);
              break;
            }
          case 'ArrowUp':
            {
              newFocusedItem = getNextSeriesFocusedItem(prevState);
              break;
            }
          default:
            break;
        }
        if (newFocusedItem !== prevState.keyboardNavigation.item) {
          event.preventDefault();
          return (0, _extends2.default)({}, prevState, prevState.highlight && {
            highlight: (0, _extends2.default)({}, prevState.highlight, {
              lastUpdate: 'keyboard'
            })
          }, prevState.interaction && {
            interaction: (0, _extends2.default)({}, prevState.interaction, {
              lastUpdate: 'keyboard'
            })
          }, {
            keyboardNavigation: (0, _extends2.default)({}, prevState.keyboardNavigation, {
              item: newFocusedItem
            })
          });
        }
        return prevState;
      });
    }
    element.addEventListener('keydown', keyboardHandler);
    element.addEventListener('blur', removeFocus);
    return () => {
      element.removeEventListener('keydown', keyboardHandler);
      element.removeEventListener('blur', removeFocus);
    };
  }, [svgRef, removeFocus, params.enableKeyboardNavigation, store]);
  (0, _useEnhancedEffect.default)(() => store.update(prev => prev.keyboardNavigation.enableKeyboardNavigation === params.enableKeyboardNavigation ? prev : (0, _extends2.default)({}, prev, {
    keyboardNavigation: (0, _extends2.default)({}, prev.keyboardNavigation, {
      enableKeyboardNavigation: !!params.enableKeyboardNavigation
    })
  })), [store, params.enableKeyboardNavigation]);
  return {};
};
exports.useChartKeyboardNavigation = useChartKeyboardNavigation;
useChartKeyboardNavigation.getInitialState = params => ({
  keyboardNavigation: {
    item: null,
    enableKeyboardNavigation: !!params.enableKeyboardNavigation
  }
});
useChartKeyboardNavigation.params = {
  enableKeyboardNavigation: true
};
